librz/reg/profile.c

```diff
@@ -1,245 +1,452 @@
+		RZ_LOG_WARN("Ignoring duplicated register definition '%s'.\n", name);
+		return true;
+	}
+
+	RzRegProfileDef *def = RZ_NEW0(RzRegProfileDef);
+	if (!def) {
+		RZ_LOG_WARN("Unable to allocate memory.\n");
+		return false;
+	}
+	def->name = strdup(name);
+
+	if (!parse_type(def, rz_list_get_n(tokens, 0))) {
+		RZ_LOG_WARN("Invalid register type.\n");
+		goto reg_parse_error;
+	}
+
+	def->size = parse_size(rz_list_get_n(tokens, 2));
+	if (def->size == UT32_MAX || def->size == 0) {
+		RZ_LOG_WARN("Invalid register size.\n");
+		goto reg_parse_error;
+	}
+
+	def->packed = parse_size(rz_list_get_n(tokens, 4)) * 8;
+	if (def->packed == UT32_MAX) {
+		RZ_LOG_WARN("Invalid register packed size.\n");
+		goto reg_parse_error;
+	}
+
+	if (!parse_offset(rz_list_get_n(tokens, 3), def)) {
+		RZ_LOG_WARN("Invalid register offset.\n");
+		goto reg_parse_error;
+	}
 
-	// This is optional
-	if (n == 6) {
-		if (*tok[5] == '#') {
+	// Comments and flags are optional
+	if (rz_list_length(tokens) == 6) {
+		const char *comment_flag = rz_list_get_n(tokens, 5);
+		rz_return_val_if_fail(comment_flag, false);
+		if (comment_flag[0] == '#') {
 			// Remove # from the comment
-			item->comment = strdup(tok[5] + 1);
+			def->comment = strdup(comment_flag + 1);
 		} else {
-			item->flags = strdup(tok[5]);
+			def->flags = strdup(comment_flag);
 		}
 	}
 
-	item->arena = type2;
-	if (!reg->regset[type2].regs) {
-		reg->regset[type2].regs = rz_list_newf((RzListFree)rz_reg_item_free);
-	}
-	rz_list_append(reg->regset[type2].regs, item);
-	if (!reg->regset[type2].ht_regs) {
-		reg->regset[type2].ht_regs = ht_pp_new0();
+	rz_list_append(reg->reg_profile.defs, def);
+
+	return true;
+
+reg_parse_error:
+	rz_reg_profile_def_free(def);
+	return false;
+}
+
+/**
+ * \brief Parses a register profile string. Each line is either a register alias or a register definiton.
+ *
+ * A register alias string is of the following form:
+ * "=<alias>  <name>\n"
+ *
+ * A register definition string is of the following form:
+ * "(<sub-type>@)main-type  <name>  .<size>  <byte offset>(.<bit offset>)  <packed>  (# <comment> OR <flags>)\n"
+ *
+ * Elements in "()" are optional.
+ * Each "<...>" token is separated by a tab character.
+ *
+ * * alias: Register alias (e.g. PC, A1 etc.)
+ * * name: Register name.
+ * * size: Register size in bits.
+ * * main-type: Register type: gpr, fpr, ctr, flg etc.
+ * * sub-type: The second register type (e.g. xmm@fpu : xmm is sub-type of fpu)
+ * * byte offset: Offset into register profile in bytes.
+ * * bit offset: Offset into the byte offset in bits.
+ * * packed: Packed size of the register in bytes.
+ * * comment: A comment about the register.
+ * * Flags this register holds.
+ *
+ * \param reg Register struct which holds all register items.
+ * \param profile Register profile string.
+ * \return false On failure.
+ * \return true On success.
+ */
+static bool parse_reg_profile_str(RZ_BORROW RzReg *reg, const char *profile) {
```

Comment:
Better make this parse only the profile without touching any `RzReg` and only later apply it, something like:
```
static bool parse_reg_profile_str(RZ_OUT RzRegProfile *profile, const char *str) {
```